### ISSUE TITLE
False negative of assigning union of plain tuples to any type

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#3592]
## Environment data

-   Language Server version: v2022.6.21 Pre-Release
-   OS and version: Windows 7
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.12 Anaconda

## Code Snippet

```python
from __future__ import annotations

from typing import Any, NamedTuple

ANY_VALUE: Any = globals()[""]

class T(NamedTuple):
    v: Any  # Type here doesn't matter.

t: T | None = None

# Condition can be anything other than boolean literals.
# Type of either value A or value B needs to be `Any`, but not both.
if ANY_VALUE:  # Condition
    t_ = (
        0,  # Value A
    )
else:
    t_ = (
        ANY_VALUE,  # Value B
    )

t = t_  # ???
```

## Expected behavior

Incompatibility reported for last line.

## Actual behavior

No errors/warnings.

**Comment by Azureblade3808** [comment]
It seems nothing about named tuple, but only union of tuples with Any.

```python
from __future__ import annotations

from typing import Any

a: tuple[Any] | tuple[int] = globals()[""]
b: int = a  # ???
```

Any element type in any sub type of the union being `Any` can trigger this behavior, as long as every sub type of the union is tuple and can have a common length (including those declared with `...`).

**Comment by erictraut** [comment]
Thanks for the bug report. I understand the problem. This will be fixed in the next release.

There was a failure in a heuristic within the type validation logic for unions. If a union includes two types where one type is a proper subtype of the other, the wider of the two types can be ignored for purposes of type checking, even in cases where invariance is required. The problem here is that `tuple[Any]` is a subtype of `tuple[int]` and vice versa, so this heuristic was eliminating both types from the union during type validation. The fix is to apply this heuristic only when the subtype relationship is unidirectional.

**Comment by erictraut** [comment]
This is included in pyright 1.1.255, which I just published. It will also be included in the next release of pylance.
